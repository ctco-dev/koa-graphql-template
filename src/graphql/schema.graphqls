# Root _Query_ definition
type Query {
  # Get _Joke_ by category
  jokeByCategory(category: JokeCategoryEnum): Joke
  # Get my _LifeStory_
  myLifeStory: String!
}

# _Joke_ definition
type Joke {
  id: String!
  text: String!
  categories: [JokeCategoryEnum]
}

# Available joke's categories
enum JokeCategoryEnum {
  NERDY
  EXPLICIT
}
