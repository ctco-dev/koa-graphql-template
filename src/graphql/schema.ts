import * as fs from 'fs';
import * as path from 'path';
import { makeExecutableSchema } from 'graphql-tools';

import { schema as jokesSchema, resolvers as jokesResolvers } from './jokes';


const rootSchema: string = fs.readFileSync(path.join(__dirname, './query.graphqls'), 'utf8');

export default makeExecutableSchema({
  typeDefs: [rootSchema, jokesSchema],
  resolvers: { ...jokesResolvers },
});
