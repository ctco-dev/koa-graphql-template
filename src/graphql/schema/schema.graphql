# import * from "nodeInterface"
# import * from "pageInfoType"

# import Query.* from "jokes.graphql"
# import Query.* from "swapi"
