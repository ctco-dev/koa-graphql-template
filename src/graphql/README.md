GraphQL
=======

Currently the best approach seems to keep all schema definitions in a single file
 `schema.graphqls`. If it increases and becomes huge, probably it could be split into
 sensible parts and placed somewhere like `schemaDefs`. For merging schema pieces
 see [Modularizing the schema](http://dev.apollodata.com/tools/graphql-tools/generate-schema.html#modularizing).
 