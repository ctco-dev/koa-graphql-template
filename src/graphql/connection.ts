import { connectionDefinitions } from 'graphql-relay-tools';

/**
  * Constructs a GraphQL connection field config; it is assumed
  * that the object has a property named `prop`, and that property
  * contains a list of types.
  */
export function connectTypes(name: string, prop: string, type: string) {
  const { connectionType } = connectionDefinitions({
    name,
    nodeType: type,
    connectionFields: `
      totalCount: Int
      ${prop}: [${type}]
    `,
  });

  return connectionType;
}
