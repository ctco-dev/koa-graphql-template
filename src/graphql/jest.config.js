module.exports = Object.assign({}, require('../../jest.default.config'), {
  "setupFiles": [
    "./__tests__/__setup__/env-test.ts"
  ],
});
