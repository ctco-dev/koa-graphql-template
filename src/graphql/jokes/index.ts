import * as fs from 'fs';
import * as path from 'path';

import resolvers from './resolvers';
import * as types from './types';

const schema: string = fs.readFileSync(path.join(__dirname, './jokes.graphqls'), 'utf8');

export {
  schema,
  resolvers,
  types,
};
