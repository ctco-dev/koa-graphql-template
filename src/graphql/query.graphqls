# _Joke_ definition
type Joke {
  id: String!
  text: String!
  categories: [JokeCategoryEnum]
}

# Available joke's categories
enum JokeCategoryEnum {
  NERDY
  EXPLICIT
}

# Root _Query_ definition
type Query {
  # Get _Joke_ by category
  jokeByCategory(category: JokeCategoryEnum): Joke
}
