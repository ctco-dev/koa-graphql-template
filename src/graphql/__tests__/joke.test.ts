import { graphql } from 'graphql';
import schema from '../schema';
import * as nock from 'nock';

const rootValue = {};
const context = {};

beforeAll(() => {
  const response = {
    type: 'success',
    value: {
      id: 1000,
      joke: "Mock Chuck Norris's database has only one table, 'Kick', which he DROPs frequently.",
      categories: [
        'nerdy'
      ]
    }
  };

  // mock service endpoint
  nock(process.env.JOKE_SERVICE_URI)
    .get('/jokes/random')
    .query({ limitTo: '[nerdy]' })
    .reply(200, response);
});

describe('query.joke', () => {
  it('should match snapshot', async () => {
    const query = `
      query Q {
        joke {
          text
          category
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context);
    const { data } = result;

    expect(data).toMatchSnapshot();
  });
});

