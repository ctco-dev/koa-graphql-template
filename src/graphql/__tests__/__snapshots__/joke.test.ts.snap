// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query.joke should match snapshot 1`] = `
Object {
  "joke": Object {
    "category": "NERDY",
    "text": "Mock Chuck Norris's database has only one table, 'Kick', which he DROPs frequently.",
  },
}
`;
